File: taurus-protect-sdk-go/pkg/protect/mapper/webhook.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// WebhookFromDTO converts an OpenAPI Webhook to a domain Webhook.
func WebhookFromDTO(dto *openapi.TgvalidatordWebhook) *model.Webhook {
	if dto == nil {
		return nil
	}

	webhook := &model.Webhook{
		ID:     safeString(dto.Id),
		Type:   safeString(dto.Type),
		URL:    safeString(dto.Url),
		Status: safeString(dto.Status),
	}

	// Convert timestamps
	if dto.TimeoutUntil != nil {
		webhook.TimeoutUntil = *dto.TimeoutUntil
	}
	if dto.CreatedAt != nil {
		webhook.CreatedAt = *dto.CreatedAt
	}
	if dto.UpdatedAt != nil {
		webhook.UpdatedAt = *dto.UpdatedAt
	}

	return webhook
}

// WebhooksFromDTO converts a slice of OpenAPI Webhooks to domain Webhooks.
func WebhooksFromDTO(dtos []openapi.TgvalidatordWebhook) []*model.Webhook {
	if dtos == nil {
		return nil
	}
	webhooks := make([]*model.Webhook, len(dtos))
	for i := range dtos {
		webhooks[i] = WebhookFromDTO(&dtos[i])
	}
	return webhooks
}

File: taurus-protect-sdk-typescript/src/mappers/webhook.ts
/**
 * Webhook mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { Webhook } from '../models/webhook';
import { WebhookStatus } from '../models/webhook';
import { safeDate, safeInt, safeMap, safeString } from './base';

/**
 * Maps a webhook status string to WebhookStatus enum.
 */
function mapWebhookStatus(status: string | undefined): WebhookStatus | undefined {
  if (!status) {
    return undefined;
  }
  const upper = status.toUpperCase();
  switch (upper) {
    case 'ACTIVE':
    case 'ENABLED':
      return WebhookStatus.ACTIVE;
    case 'INACTIVE':
    case 'DISABLED':
      return WebhookStatus.INACTIVE;
    case 'FAILED':
      return WebhookStatus.FAILED;
    default:
      return undefined;
  }
}

/**
 * Maps a webhook DTO to a Webhook domain model.
 */
export function webhookFromDto(dto: unknown): Webhook | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const eventsRaw = d.events ?? d.type ?? d.types;
  let events: string[] | undefined;

  if (Array.isArray(eventsRaw)) {
    events = eventsRaw.map((e) => String(e));
  } else if (typeof eventsRaw === 'string') {
    // API may return comma-separated event types
    events = eventsRaw.split(',').map((e) => e.trim());
  }

  return {
    id: safeString(d.id ?? d.webhookId ?? d.webhook_id),
    url: safeString(d.url),
    events,
    status: mapWebhookStatus(safeString(d.status)),
    secret: safeString(d.secret),
    createdAt: safeDate(d.createdAt ?? d.created_at ?? d.creationDate),
    updatedAt: safeDate(d.updatedAt ?? d.updated_at ?? d.modificationDate),
    failureCount: safeInt(d.failureCount ?? d.failure_count ?? d.consecutiveFailures),
    lastFailureMessage: safeString(d.lastFailureMessage ?? d.last_failure_message ?? d.lastError),
  };
}

/**
 * Maps an array of webhook DTOs to Webhook domain models.
 */
export function webhooksFromDto(dtos: unknown[] | null | undefined): Webhook[] {
  return safeMap(dtos, webhookFromDto);
}
